src/resources/base.js

```diff
@@ -0,0 +1,171 @@
+import ValidationError from '../errors/validation';
+import NotImplementedError from '../errors/notImplemented';
+
+export default api => (
+  class Base {
+    static url = null;
+    static _name = null;
+    static propTypes = {};
+    static jsonIdKeys = [];
+
+    static async retrieve(id) {
+      try {
+        const res = (await api.get(`${this.url}/${id}`));
+        return this.create(res.body);
+      } catch (e) {
+        return Promise.reject(e);
+      }
+    }
+
+    static async all(query = {}) {
+      try {
+        const res = await api.get(`${this.url}`, { query });
+        return this.unwrapAll(res.body).map(this.create.bind(this));
+      } catch (e) {
+        return Promise.reject(e);
+      }
+    }
+
+    static notImplemented(fnName) {
+      return Promise.reject(new NotImplementedError(fnName, this.url));
+    }
+
+    static wrapJSON(json) {
+      return { [this._name.toLowerCase()]: json };
+    }
+
+    static unwrap(data) {
+      return data;
+    }
+
+    static create(data) {
+      return new this(this.unwrap(data));
+    }
+
+    static unwrapAll(data) {
+      return data;
+    }
+
+    _validationErrors = null;
+
+    // suppressVAlidation is used when creating objects from API responses-
+    // the API returns keys that we don't later on use for creating or editing.
+    // We want access to read these, but we don't want to throw validation
+    // errors; they're valid, but read-only. Note to self: maybe add a readonly
+    // type that uses getters and setters?
+    constructor(data = {}) {
+      this.mapProps(data);
+    }
+
+    validateProperties(throwOnFailure = true) {
```

Comment:
It's reading its properties as an instance, so it shouldn't be static (I suppose it _could_ be static and have the proptypes and properties passed in, but I don't see a benefit.)